### PR TITLE
Replace use of sweet_pickle in naming, and delete sweet_pickle subpackage

### LINKED ISSUE #133: Fix multiple versioned pickle wrappers
Apptools provides the `sweet_pickle` subpackage to handle versioning of pickled objects to allow unpickling of objects after changes to the class structure.

The `persistence` subpackage has a slightly different focus: it is more about extracting state as a dictionary, but it has similar themes.

Mayavi uses `persistence`, but it also uses `apptools.naming` which internally uses `sweet_pickle`, so we cannot simply deprecate `sweet_pickle`.

### PR DESCRIPTION
**Pull request by aaronayres35**
fixes #133 

This PR very possibly is biting the bullet too early on this, so I am happy to hold off on merging or close if need be. (we might not want to kill `sweet_pickle`!)

`sweet_pickle` was previously only used in a very minor way in `apptools.naming`, and was not used anywhere else in `apptools`.  Additionally, AFAIK it is not used in other projects currently (this might not be the case in which case we can simply close this PR). This PR replaces its use in `apptools.naming` and deletes the `sweet_pickle` subpackage.  

For more detailed commentary on this see #133 and specifically these comments: https://github.com/enthought/apptools/issues/133#issuecomment-724917134 https://github.com/enthought/apptools/issues/133#issuecomment-724933060

It was also mentioned in the EEP that in `envisage` "The `single_project` plugin uses `sweet_pickle` and `naming`.".  However, `single_project` was removed by https://github.com/enthought/envisage/pull/331

**Checklist**
- [x] Add a news fragment if this PR is news-worthy for end users. (see docs/releases/README.rst)

Also, we could do a quick scream test on slack like we did for `apptools.lru_cache`.

### CODE REVIEW

**Review by kitchoi — COMMENTED** [review]
Persistence is tricky because we also need to think about files that have been created in the past. Although pickling is not meant for long-term persistence, we still need to think about them...

Looks like Mayavi uses of apptools.naming (which uses `sweet_pickle`) is mostly transient: It is about binding variables to the Python shell.

As for the changes in `apptools.naming.object_serializer`, it seems those lines are indirectly covered by tests exercising `apptools.naming.pyfs_context` or `apptools.naming.context._bind`, and I can't find a test to ensure that round-tripping is working correctly, i.e. what is saved by `ObjectSerializer.save` can be loaded by `ObjectSerializer.load` correctly.

So I added the following test, which passes on master:

<details>
    <summary> New test for roundtripping </summary>

```
# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
# All rights reserved.
#
# This software is provided without warranty under the terms of the BSD
# license included in LICENSE.txt and may be redistributed only under
# the conditions described in the aforementioned license. The license
# is also available online at http://www.enthought.com/licenses/BSD.txt
#
# Thanks for using Enthought open source!
import os
import shutil
import tempfile
import unittest

from traits.api import cached_property, HasTraits, Property, Str, Event

from apptools.naming.api import ObjectSerializer


class FooWithTraits(HasTraits):
    """ Dummy HasTraits class for testing ObjectSerizalizer."""

    full_name = Str()

    last_name = Property(depends_on="full_name")

    event = Event()

    @cached_property
    def _get_last_name(self):
        return self.full_name.split(" ")[-1]

class TestObjectSerializer(unittest.TestCase):

    def setUp(self):
        self.tmpdir = tempfile.mkdtemp()
        self.addCleanup(shutil.rmtree, self.tmpdir)
        self.tmp_file = os.path.join(self.tmpdir, "tmp.pickle")

    def test_save_load_roundtrip(self):
        # Test HasTraits objects can be serialized and deserialized as expected
        obj = FooWithTraits(full_name="John Doe")

        serializer = ObjectSerializer()
        serializer.save(self.tmp_file, obj)

        self.assertTrue(serializer.can_load(self.tmp_file))
        deserialized = serializer.load(self.tmp_file)

        self.assertIsInstance(deserialized, FooWithTraits)
        self.assertEqual(deserialized.full_name, "John Doe")
        self.assertEqual(deserialized.last_name, "Doe")
```

</details>

But it fails on this branch:
```
======================================================================
ERROR: test_save_load_roundtrip (apptools.naming.tests.test_object_serializer.TestObjectSerializer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/apptools/apptools/naming/tests/test_object_serializer.py", line 48, in test_save_load_roundtrip
    deserialized = serializer.load(self.tmp_file)
  File "/Users/kchoi/Work/ETS/apptools/apptools/naming/object_serializer.py", line 58, in load
    obj = VersionedUnpickler(f).load()
  File "/Users/kchoi/Work/ETS/apptools/apptools/persistence/versioned_unpickler.py", line 37, in load
    ret = Unpickler.load(self)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/pickle.py", line 1210, in load
    dispatch[key[0]](self)
TypeError: 'classmethod' object is not callable
```

Something still needs fixing?

**Comment by aaronayres35** [comment]
Note there is this integration test for VersionedUnpickler: https://github.com/enthought/apptools/blob/master/integrationtests/persistence/test_persistence.py

(I'm simply linking this because I saw it existed, it is very possibly irrelevant / we may just want to delete it)

File: apptools/naming/object_serializer.py [hunk 1]
```diff
@@ -86,9 +83,7 @@ def save(self, path, obj):
         # Pickle the object.
         f = open(actual_path, 'wb')
         try:
-            sweet_pickle.dump(obj, f, 1)
-#            cPickle.dump(obj, f, 1)
-#            pickle.dump(obj, f, 1)
+            pickle.dump(obj, f, 1)
```
[the remark this answers is not in the document]
Comment (aaronayres35) [reply apptools/naming/object_serializer.py:86]: Although previously it was `sweet_pickle.load(f)` and `sweet_pickle.dump(obj, f, 1)` making them appear in sync, actual implementations of `sweet_pickle.load` and `sweet_pickle.dump` are effectively the same as they are here.  Ie. `dump` just calls `pickle.dump` and `load` actually comes from the `VersionedUnpickler` class (as mentioned https://github.com/enthought/apptools/issues/133#issuecomment-724917134)

However, I do agree it is more readable to have them be consistent and we could easily add a dumb `dump` method to the `persistence` `VersionedUnpickler` class

**Comment by aaronayres35** [comment]
> So I added the following test, which passes on master:
> 
>  New test for roundtripping
> But it fails on this branch:
> 
> ```
> ======================================================================
> ERROR: test_save_load_roundtrip (apptools.naming.tests.test_object_serializer.TestObjectSerializer)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/kchoi/Work/ETS/apptools/apptools/naming/tests/test_object_serializer.py", line 48, in test_save_load_roundtrip
>     deserialized = serializer.load(self.tmp_file)
>   File "/Users/kchoi/Work/ETS/apptools/apptools/naming/object_serializer.py", line 58, in load
>     obj = VersionedUnpickler(f).load()
>   File "/Users/kchoi/Work/ETS/apptools/apptools/persistence/versioned_unpickler.py", line 37, in load
>     ret = Unpickler.load(self)
>   File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/pickle.py", line 1210, in load
>     dispatch[key[0]](self)
> TypeError: 'classmethod' object is not callable
> ```
> 
> Something still needs fixing?

So, I was able to get this test (and all the tests) to pass on this branch, by simply removing the `@classmethod` decorator I had previously added to the `load_build` method on `VersionedUnpickler`.  I had originally added it to match the `VersionedUnpickler` from `sweet_pickle` and to follow what was said in the comment:
https://github.com/enthought/apptools/blob/a693c79e8547ccaafcf89f230c0b17edbc780655/apptools/persistence/versioned_unpickler.py#L93-L104

It appears this code was copy pasted from `sweet_pickle` and then this decorator likely got removed because it was causing bugs somehow?  In any case I am still quite confused as your test passes on master, where in `sweet_pickle` that `@classmethod` decorator is there, however, adding it here makes that test fail.  I need to investigate more

File: apptools/persistence/versioned_unpickler.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 # Standard library imports
-from pickle import *
+from pickle import _Unpickler as Unpickler
```
[the remark this answers is not in the document]
Comment (aaronayres35) [reply apptools/persistence/versioned_unpickler.py:2]: I think previously this may have been untested? `pickle.Unpickler` has no attribute `dispatch` so here I immediately see test failures with:
```
======================================================================
ERROR: test_save_load_roundtrip (apptools.persistence.tests.test_roundtrip.TestObjectSerializer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/aayres/Desktop/apptools/apptools/persistence/tests/test_roundtrip.py", line 48, in test_save_load_roundtrip
    deserialized = serializer.load(self.tmp_file)
  File "/Users/aayres/Desktop/apptools/apptools/naming/object_serializer.py", line 58, in load
    obj = VersionedUnpickler(f).load()
  File "/Users/aayres/Desktop/apptools/apptools/persistence/versioned_unpickler.py", line 33, in load
    dispatch = self.dispatch
AttributeError: 'VersionedUnpickler' object has no attribute 'dispatch'

----------------------------------------------------------------------
```

`pickle._Unpickler` does have `dispatch`, see here: https://github.com/python/cpython/blob/aa01011003bb855cd52abfd49f2443446590d913/Lib/pickle.py#L1224

On master no tests fail (including your new one), because the logic is coming from the `VersionedUnpickler` object from `sweet_pickle` which subclasses `pickle._Unpickler` (note the classes are almost entirely copy and pasted otherwise).  



So, in summary, if we want to go on importing `pickle.Unpickler` instead of `pickle._Unpickler`, I believe we would have to rewrite the code to not use `dispatch`.  There very well be a way to do that seamlessly, it is just another change we would need to make.  The way the code is written now, it seems to assume we are using `pickle._Unpickler`?

File: apptools/naming/object_serializer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kitchoi) [reply apptools/naming/object_serializer.py:88]: Opened https://github.com/enthought/apptools/issues/231

File: apptools/persistence/versioned_unpickler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kitchoi) [reply apptools/persistence/versioned_unpickler.py:2]: I see. Thanks.

Now I see that in Python 2.7, there is no such `pickle._Unpickler`, the same class was named `Unpickler` then. However in the documentation, it is clear that `dispatch` is not part of the public API, so `NewUnpickler` has depended on private and internal implementation details... which is a risky thing to do. 

For this PR should maintain behaviour of `persistence`, we'd have to use `_Unpickler` but it is clear that `dispatch` should not be depended on. Please could you open an issue to remove the dependency on private interface and reference the issue in a comment here?

**Comment by kitchoi** [comment]
> It appears this code was copy pasted from sweet_pickle and then this decorator likely got removed because it was causing bugs somehow? In any case I am still quite confused as your test passes on master, where in sweet_pickle that @classmethod decorator is there, however, adding it here makes that test fail. I need to investigate more

I think that's because this line wasn't removed after `@classmethod` was added:
https://github.com/enthought/apptools/blob/6de956b3c683b01739d13f56d3b4ab70649bfb4d/apptools/persistence/versioned_unpickler.py#L96

This line is equivalent to adding `@classmethod` to `def load_build`, but we don't need both.

File: apptools/persistence/versioned_unpickler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaronayres35) [reply apptools/persistence/versioned_unpickler.py:2]: opened #232

**Comment by kitchoi** [comment]
@aaronayres35 The roundtripping test above is quite useful for making sure the changes in this PR has not broken object serializer (in the most trivial ways at least). I think it would be good to add the test as part of this PR, it will also serve as future reference as to what extent the object serializer remains to work.

I saw that you have removed `apptools/sweet_pickle/tests/__init__.py`. While the tests in that folder are not discovered and run by `unittest discover` any more, the test files are still in the folders.  It would be a strong proof to the claim "users of sweet_pickle can transition to using ``apptools.persistance`` and pickle from the python standard library" if we can convert the existing tests in `sweet_pickle` to use `persistence`, what do you think?

**Comment by aaronayres35** [comment]
> @aaronayres35 The roundtripping test above is quite useful for making sure the changes in this PR has not broken object serializer (in the most trivial ways at least). I think it would be good to add the test as part of this PR, it will also serve as future reference as to what extent the object serializer remains to work.

Yeah that makes sense, I will add the test to this PR and have it sit in `naming` as it is intended for `ObjectSerializer`
 
> I saw that you have removed `apptools/sweet_pickle/tests/__init__.py`. While the tests in that folder are not discovered and run by `unittest discover` any more, the test files are still in the folders. It would be a strong proof to the claim "users of sweet_pickle can transition to using `apptools.persistance` and pickle from the python standard library" if we can convert the existing tests in `sweet_pickle` to use `persistence`, what do you think?

Somehow when merging / addressing conflicts I must have made a mistake and readded `sweet_pickle` files.  I had intended to remove the entire sub package in this PR.  What you said does make sense though as far as supporting the "users of sweet_pickle can transition to using `apptools.persistance` and pickle from the python standard library" claim. 
But in that case where should these tests sit?  Should we leave a basically empty `sweet_pickle` package keeping the old tests?  Should I move the tests into `persistence` after they have been modified (I think that would make sense)

**Comment by kitchoi** [comment]
> Somehow when merging / addressing conflicts I must have made a mistake and readded sweet_pickle files. I had intended to remove the entire sub package in this PR. 

I see! That's probably why I did not see them when I checked out the branch and now they reappear.

> Should I move the tests into persistence after they have been modified (I think that would make sense)

Adding them to `persistence` would make sense to me too.

**Comment by aaronayres35** [comment]
Note `sweet_pickle` is also used in `codetools`, but all that is used is `sweet_pickle.load` and `sweet_pickle.dump` which can be replaced as is done here in `apptools.naming`.

**Comment by aaronayres35** [comment]
> I saw that you have removed `apptools/sweet_pickle/tests/__init__.py`. While the tests in that folder are not discovered and run by `unittest discover` any more, the test files are still in the folders. It would be a strong proof to the claim "users of sweet_pickle can transition to using `apptools.persistance` and pickle from the python standard library" if we can convert the existing tests in `sweet_pickle` to use `persistence`, what do you think?

Some of the `sweet_pickle` tests are a bit `sweet_pickle` specific.  For example, the tests in 
https://github.com/enthought/apptools/blob/master/apptools/sweet_pickle/tests/test_updater.py
basically each just test each specific method on the `Updater` class.  `persistence` also has an `Updater` class but it is definitely not the same.  It seems they go about doing things in different ways.  Some of these tests I'm not sure how to directly modify to use `persistence`.  

In fact, I am not confident that `persistence` can replace all of the functionality provided by `sweet_pickle`.  I am confident thought that it can replace all the functionality of `sweet_pickle` that I have seen _used_.  Which at this point all I have really seen is calls to `sweet_pickle.load`, `sweet_pickle.loads`, `sweet_pickle.dump`, and `sweet_pickle.dumps`, which can be replaced like they are in this PR in `apptools.naming`

**Comment by kitchoi** [comment]
> In fact, I am not confident that persistence can replace all of the functionality provided by sweet_pickle. I am confident thought that it can replace all the functionality of sweet_pickle that I have seen used. Which at this point all I have really seen is calls to sweet_pickle.load, sweet_pickle.loads, sweet_pickle.dump, and sweet_pickle.dumps, which can be replaced like they are in this PR in apptools.naming

I see. Let's try to convert the tests that can be converted. Then we can more confidently say which functionality can be migrated and which are being discarded.

File: apptools/sweet_pickle/tests/test_class_mapping.py [hunk 3]
```diff
@@ -1,99 +0,0 @@
...
-        self.registry.add_mapping_to_class(Baz.__module__, Baz.__name__, Foo)
-
-        # Validate that an exception is raised when trying to unpickle an
-        # instance anywhere within the circular definition.
-        def fn(o):
-            sweet_pickle.loads(sweet_pickle.dumps(o))
-
-        self.assertRaises(sweet_pickle.UnpicklingError, fn, Foo())
-        self.assertRaises(sweet_pickle.UnpicklingError, fn, Bar())
-        self.assertRaises(sweet_pickle.UnpicklingError, fn, Baz())
```
Comment (aaronayres35) [inline apptools/sweet_pickle/tests/test_class_mapping.py:81]: I believe cycles aren't possible with how `persistence` is set up, so I deleted this test.  The following test I modified slightly as there is only one transition that can occur.  ie with persistence you can't go Foo->Bar->Baz in one roundtrip

File: apptools/persistence/tests/test_state_function.py [hunk 4]
```diff
@@ -89,47 +82,25 @@ def test_normal_setstate(self):
...
         start = Bar()
-        end = sweet_pickle.loads(sweet_pickle.dumps(start))
-        self.assertIsInstance(end, Baz)
-        self._assertAttributes(end, 2, None)
-        self._assertAttributes(end, 3, (True, 2, 2, "bar"))
-
-    def test_unpickled_chain_functionality(self):
-        """Validates that the registered state functions are used when
-        unpickling.
-        """
```
Comment (aaronayres35) [inline apptools/persistence/tests/test_state_function.py:116]: I believe it may be possible to modify this test, to get a similar flavor of test for `apptools.persistence` by using the `setstates` attribute on an `Updater` instance that is passed to a `VersionedUnpickler`.  Unfortunately I was unable to get it to work atm.

**Comment by aaronayres35** [comment]
Orthogonal to this PR, but I think apptools could really benefit from improved docstrings (numpy style).  Obviously documentation could be improved in general, but I find myself spending a lot of time just trying to understand what arguments are, etc.  For example when workng on this PR I noticed that the `apptools.persistence.updater.Updater` class definition never mentions a `setstates` attribute.  See here:
https://github.com/enthought/apptools/blob/d794275303325de30746e1d21bbd6e4327d3e9ae/apptools/persistence/updater.py#L9-L36
It is however used here:
https://github.com/enthought/apptools/blob/d794275303325de30746e1d21bbd6e4327d3e9ae/apptools/persistence/versioned_unpickler.py#L157

 I discovered it here:
https://github.com/enthought/apptools/blob/d794275303325de30746e1d21bbd6e4327d3e9ae/integrationtests/persistence/update1.py#L39-L41

docstrings would save a lot of time spent trying to follow the code

File: apptools/persistence/tests/test_two_stage_unpickler.py [hunk 5]
```diff
@@ -79,24 +82,19 @@ def test_generic(self):
...
-            print("\ta.x: %s" % new_a.x)
-            print("\ta.b_ref.y: %s" % new_a.b_ref.y)
-        except Exception as msg:
-            print("\t%s" % "Expected Error".center(75, "*"))
-            print("\t%s" % msg)
-            print("\t%s" % ("*" * 75))
+            new_a.x
+            new_a.b_ref.y
+        except Exception:
+            pass
```
Comment (kitchoi) [inline apptools/persistence/tests/test_two_stage_unpickler.py:88]: Only the second line fails on master:
```
    new_a.b_ref.y
AttributeError: 'B' object has no attribute 'y'
```

While I am not sure whether it was intended that `new_a.x` can be accessed, the test is expecting an exception somewhere from these two lines. In the spirit of keeping the behaviour we know now:
```
# Accessing new_a.x is okay
new_a.x
# Accessing y directly would fail
with self.assertRaisesRegex(
        AttributeError, "'B' object has no attribute 'y'"):
    new_a.b_ref.y
```
(Different styles for checking the exception message exist.)

File: apptools/persistence/tests/test_state_function.py [hunk 6]
```diff
@@ -7,53 +7,53 @@
...
-# multiple names in sys.modules due to relative naming. Nice.
-from apptools.sweet_pickle.tests.state_function_classes import Foo, Bar, Baz
-
-##############################################################################
-# State functions to use within the tests
-##############################################################################
-
-
-def bar_state_function(state):
+def bar_state_function(self, state):
```
Comment (kitchoi) [inline apptools/persistence/tests/test_state_function.py:30]: Looks like this function is unused now?

File: apptools/persistence/tests/test_state_function.py (hunk 4 above)
Comment (kitchoi) [reply apptools/persistence/tests/test_state_function.py:116]: I looked into this briefly... IMO, it looks evil! I opened https://github.com/enthought/apptools/issues/238.

**Review by kitchoi — COMMENTED** [review]
Thank you very much for converting the tests, it really helps us to understand what can or cannot be migrated!
Some suggestions for the tests and the news fragment given the new information. The rest looks good to me.

**Review by kitchoi — APPROVED** [review]
LGTM
